session: restore same :terminal buf split windows

AlexandreDubray opened the pull request:
Attempt to fix for issue #5250 .

When creating a terminal buffer with command like `:edit term://.//16450:/bin/bash`, the buffer get a different name (depending on the PID of the terminal instance). Thus, further call to `bufexists('term://.//16450:/bin/bash)` will return false.

To fix this, I added a `char_u *` field in the structure `Terminal` so when looking for a buffer by name, it look at this name too.

There is still one test that need to pass at the moment.

Commit:
mksession: restore same :term buf in split windows

Problem: When session-restore creates a terminal buffer with command
         like `:edit term://.//16450:/bin/bash`, the buffer gets
         a different name (depends on PID). Thus the later call to
         `bufexists('term://.//16450:/bin/bash)` will return false.
Solution: Force the buffer name with :file. This as least ensures
         the same buffer will show in multiple windows correctly, as
         expected when saving the session.  But it still has problems:
           1. the PID in the buffer name is bogus
           2. redundant :terminal buffers still hang around

fix #5250

justinmk commented:
> When creating a terminal buffer with command like :edit term://.//16450:/bin/bash, the buffer get a different name (depending on the PID of the terminal instance). 

But the buffer name can be renamed _back_ to `term://.//16450:/bin/bash` immediately after the `:edit`, via `:file`:

    :file term://.//16450:/bin/bash

Then `bufexists('term://.//16450:/bin/bash') ` returns `1`.

Though it's a bit ugly, it is _much_ better than adding a bunch of special-case code + field name.  Instead, the ugly part lives only in session-restore code (which is mostly VimL).

And the buffer name with the new, correct, PID can be restored _again_ with `:file` after the session has been loaded.

justinmk reviewed (changes requested):
see comment

grebe commented:
Is the correct thing to do renaming the buffer to have an incorrect PID? I played around with making the session restore viml refer to buffers by number rather than by name (i.e. the `if bufexists` stuff)

justinmk commented:
> Is the correct thing to do renaming the buffer to have an incorrect PID? 

It's only temporary: after the session is loaded they can be renamed again to reflect the real (new) PID. And it's better than lots of special cases throughout the core.

> I played around with making the session restore viml refer to buffers by number rather than by name 

I wouldn't bother with that, it will be extremely fragile.  Buffers can be created at any time, the user could have a custom `SessionLoadPost` handler, or whatever. As soon as an unexpected buffer is created the old numbers will be invalid.

AlexandreDubray commented:
I modified the code as you suggested (which is indeed much better !). To restore the name of the buffers, I used an array to store the mapping, I don't know if there is a prettier way to do it.

I still need to add test for this.

ZyX-I commented on src/nvim/ex_docmd.c:
```diff
@@ -8973,6 +8977,15 @@ makeopens (
       next_arg_idx = wp->w_arg_idx;
     }
+		
+		if (put_line(fd,"for pair in buffer_names") == FAIL
```
Why do you use `pair` and `get()`? And outright incorrect and overly verbose `execute join`: neither join is necessary at all nor you do escaping. I would write that as

```VimL
let s:buffer_names = []
…
call add(…)
…
for [s:src, s:tgt] in s:buffer_names
  if bufexists(s:src)
    execute 'buffer' fnameescape(s:src)
    execute 'file' fnameescape(s:tgt)
  endif
endfor
unlet s:buffer_names s:src s:tgt
```

(And I do not understand now where this code goes, but should not it save what is the current buffer number and do `execute 'buffer' s:saved_curbuf` afterwards?)

ZyX-I commented on src/nvim/ex_docmd.c:
```diff
@@ -9256,6 +9269,19 @@ put_view (
       }
+
+			if (fputs("call add(buffer_names, [bufname('%'),'", fd) < 0
+					|| ses_fname(fd, wp->w_buffer, flagp, false) == FAIL
```
Must not use `ses_fname` here, it does the equivalent of `fnameescape()` which is not correct for double quotes. You need something like `string()` which will also escape NL, or like `json_encode()` which will work with non-unicode strings; I know both are or may be made (second is static inline `convert_to_json_string()`) accessible from C, but do not know whether there actually is anything which may safely dump string to file.

ZyX-I replied:
Just tested, it appears that `ses_fname` is buggy also for intented purpose: can’t work with files with newlines.

AlexandreDubray replied:
Since `ses_fname` is buggy and use everywhere in code , should it be fixed for the two case or to change it everywhere?

ZyX-I replied:
@AlexandreDubray Even though it is buggy for intended purpose, intended purpose still is not valid for expressions. As it is named `ses_fname` you can overload it for expressions going into file (without renaming, I mean), but I do not think that string escaping suitable for dumping into sourced file would always be useful for sessions only, better have separate function. Or overload `convert_to_json_string`, code is going to be very similar in any case. You would need to uninline it, make not static, add boolean argument (it must not error out, dumping any possible string instead: file names with non-unicode bytes are valid, but not JSON strings) and rename to something like `encode_convert_to_double_quoted_string`.

ZyX-I replied:
Ah, and it also should not use surrogate pairs to dump characters which have too big unicode codepoint number, unlike JSON VimL has `\U01234567` for that.

AlexandreDubray replied:
@ZyX-I Oh ok, I think I misunderstood the initial problem. So what would be the best (safest) way to write the buffer name into the session file?

timofonic commented:
Why still fixing VimScript instead replacing it with C/C++ or Lua or whatever (I did read there are interesting LuaJit forks somewhere)? I'm just curious, sorry if it's an stupid question. I'm an ignorant about the code base,  but I thought the project goal is to be a modernized and improved Vim

Why using VimScript? I don't understand it, sorry for being too dumb to get it.

timofonic commented:
Why still fixing VimScript instead replacing it with C/C++ or Lua or whatever (I did read there are interesting LuaJit forks somewhere)? I'm just curious, sorry if it's an stupid question. I'm an ignorant about the code base,  but I thought the project goal is to be a modernized and improved Vim

Why using VimScript? I don't understand it, sorry for being too dumb to get it.

ZyX-I commented:
How do you imagine Vim without VimL? And we are talking about sessions here, some people still want to use one session file for Vim and Neovim.11:22, 15 января 2018 г., Timofonic <notifications@github.com>:Why still fixing VimScript instead replacing it with C/C++ or Lua or whatever (I did read there are interesting LuaJit forks somewhere)? I'm just curious, sorry if it's an stupid question. I'm an ignorant about the code base,  but I thought the project goal is to be a modernized and improved VimWhy using VimScript? I don't understand it, sorry for being too dumb to get it.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

justinmk commented:
@timofonic You are drawing a (incorrect) conclusion from an isolated change, and ignoring (or didn't look at) all the other changes. https://gitter.im/neovim/neovim is the place for such questions.

justinmk commented on test/functional/ex_cmds/mksession_spec.lua:
```diff
@@ -46,4 +46,21 @@ describe(':mksession', function()
+		local bufname = funcs.bufname('%')
+
+		command('wincmd k')
+		eq(bufname, funcs.bufname('%'))
```
Let's check `b:terminal_job_pid` instead, that will check the real PID instead of whatever happens to be the buffer name.

```lua
local expected_pid = helpers.eval('b:terminal_job_pid')
eq(1, funcs.winnr())
command('wincmd k')
eq(2, funcs.winnr())
eq(expected_pid, helpers.eval('b:terminal_job_pid'))
```

justinmk replied:
I guess buffer id (`funcs.bufnr('%')`) would work as well. We just want to use something that wasn't possibly shuffled around by the session logic itself.

AlexandreDubray replied:
Yeah it is better :-) I also used `exe nbr . "wincmd w"` to change windows to be sure to go to the right windows directly, hope it is good.

AlexandreDubray commented:
I've made changes according to the review and fixes some lint/style issues. For the problem of the `ses_fname`, since it is not a core problem here, it might be handle in another PR? (I don't mind working on it).

If all that is good, I think the PR is good.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9256,6 +9270,19 @@ put_view (
+      if (fputs("file ", fd) < 0
+          || ses_fname(fd, wp->w_buffer, flagp, false) == FAIL
+          || put_eol(fd) == FAIL) {
+        return FAIL;
```
test passes with only this change. Are the other changes actually needed?

Maybe there's some edge case, but I'm going to merge it with only this change for now. Can always revert back to the more complex version if needed.

justinmk commented:
**Note:** the `:file ...` approach here works to solve the UX issue (and is strictly better than the existing behavior), but it of course leaves redundant `:term` buffers around.

justinmk committed:
mksession: simplify generated commands

Doing ":file …" immediately after is enough to fixup the :terminal
buffer name.

ref #5250

justinmk committed:
mksession: avoid ":file …" when restoring non-terminal bufs

justinmk committed:
mksession: always write LF "\n" line-endings

- remove `MKSESSION_NL`, `mksession_nl`
- deprecate the "unix" flag of 'sessionoptions'

There is no reason to choose CRLF or LF for session files. Instead just
always write LF.

justinmk committed:
cleanup/ex_docmd.c: remove most put_line() calls

- prefer fprintf() instead of put_line()
- PUTLINE_FAIL macro to avoid some boilerplate

justinmk committed:
lint

justinmk committed:
cleanup/ex_docmd.c: remove most put_eol() calls

justinmk committed:
mksession: always unix slashes "/" for filepaths
